docs/docs/specs/v0/composer-2.md

```diff
@@ -0,0 +1,389 @@
+        }
+      }}
+    />
+  );
+};
+```
+
+It should be clear that the entire application structure is constructed based on what is passed in to `state.surfaces` and can therefore be serialized, stored in ECHO, modified by users, or otherwise altered dynamically at runtime.
+
+It's possible for the app to enter a mode where the surface-to-component binding is itself visualized in the `Tree` therefore becoming inspectable and modifiable in-app. It's possible to locate this behavior within yet another plugin (`AppInspectorPlugin` or something).
+
+`Tree` is responsible for rendering all `GraphNodes` provided by plugins, and provides a state to the global `AppState` reflecting the currently selected node(s).
+
+In order to populate the `Tree`, plugins are first asked to present their lists of children with no `parent` node. This generates the first level items in the Tree. Then, for each node ad-nauseum, plugins are asked to return more children until the tree reaches a steady state. This allows plugins to add nodes to each other's nodes.
+
+`SpacesPlugin` provides the entire graph in this example, starting with nodes for each ECHO space, followed by subnodes representing queries or specific types, followed by objects of each type.
+
+`Selection` is responsible for sensing what nodes are selected by reading the relevant area in `AppState`, obtaining a reference to the selected `GraphNodes` and the nested ECHO objects they represent, and passing them to a `<Surface />` which knows how to choose the right component to render that ECHO object with.
+
+`StackPlugin` provides a Component that knows how to visualize a given ECHO object of type `stack`, and fulfills the `main` surface of `Selection` dynamically as a result of a `Tree` selection change.
+
+`Stack` components also use `Surfaces` to obtain components from `ComposerPlugin` and `ImagePlugin` which know how to render ECHO objects of type `text` and `image` respectively.
+
+`SearchPlugin` knows how to alter the `surfaces` section of the state such that a different control is presented instead of the `Tree` with flat search results whenever a search term is present in the input box. Alternatively, the `Tree` can equally respond to changes in the search term (global state) and filter itself down accordingly.
+
+Other plugins possible:
+
+1. the **markdown** plugin - which provides a plain text editor `Composer` for the content area and fills the Tree with plain text documents from ECHO
+2. the **filesystem** plugin - which provides import / export to folders on disk
+3. the **github** plugin - which provides nodes representing github issues and assets
+
+## Example Plugins
+
+### Tree Plugin
+
+```tsx
+import { HaloButton } from '@dxos/react-shell';
+import { Tree } from './Tree';
+import { Selection } from './Selection';
+import { Plugin } from '@dxos/app-model';
+
+export const TreePlugin: Plugin = {
+  meta: {
+    id: 'TreePlugin'
+  },
+  provides: {
+    components: {
+      HaloButton,
```

Comment:
Shouldn't be here, or at least inconsistent with above `HaloPlugin.HaloButton`

Reply:
hm? unclear why. which above?

Reply:
`HaloButton` comes from the `HaloPlugin` above https://github.com/dxos/dxos/pull/3067/files/fb24b17a05d91052343d480ed7aa6d3651b12a4f#diff-66de7abfdcafaeb017540570393b52e0f5dd07846750146869ab585ebd0fd2c2R148 whereas here it's being defined in the `TreePlugin`